### PR TITLE
Expand `InputType` with new variant: Bolt12 offer

### PR DESCRIPTION
**Pull request by ok300**
This PR extends the input parsed in `sdk-common` to recognize Bolt12 offers.

### CODE REVIEW

**Comment by ok300** [comment]
@hydra-yse your older PR #598 parses the offer to more detailed fields:

https://github.com/breez/breez-sdk-greenlight/blob/ad384b106bfcee3e225625b879b153f398b6ee7f/libs/sdk-bindings/src/breez_sdk.udl#L735-L755

As that PR is still in review, I'm not sure if that structure is final (`LNOffer`, `InputType::Bolt12Offer(LNOffer offer)`).

So would you say we need those fields as defined there? If so, I can integrate them in this PR.

**Comment by dangeross** [comment]
> So would you say we need those fields as defined there? If so, I can integrate them in this PR

I think they are needed, we need to parse the offer for the user. I also think our PR struct is out of date and I would suggest bumping the `lightning`/`lightning-invoice` dependencies. The current version doesn't parse offers according to the latest spec (e.g. the description field is now optional), which I think is not compatible with Phoenix

**Comment by JssDWt** [comment]
> As that PR is still in review, I'm not sure if that structure is final (`LNOffer`, `InputType::Bolt12Offer(LNOffer offer)`).
> 
> So would you say we need those fields as defined there? If so, I can integrate them in this PR.

Also keep in mind there are invoice requests and invoices in bolt12, next to offers. We don't have to support them, but let's make the interface in such a way that we don't have to break anything to support them in the future.

File: libs/sdk-bindings/src/breez_sdk.udl [hunk 1]
```diff
@@ -734,6 +734,7 @@ dictionary RecommendedFees {
 interface InputType {
     BitcoinAddress(BitcoinAddressData address);
     Bolt11(LNInvoice invoice);
+    Bolt12(string offer);
```
Comment (roeierez) [inline libs/sdk-bindings/src/breez_sdk.udl:737]: I think we should have here the parsed LNOffer (same as we have for other types

File: libs/sdk-bindings/src/breez_sdk.udl [hunk 2]
```diff
@@ -738,10 +738,27 @@ dictionary RecommendedFees {
...
+    Amount? amount;
+    u64? absolute_expiry;
+    string? issuer;
+};
+
 [Enum]
 interface InputType {
     BitcoinAddress(BitcoinAddressData address);
     Bolt11(LNInvoice invoice);
+    Bolt12(LNOffer offer);
```
Comment (JssDWt) [inline libs/sdk-bindings/src/breez_sdk.udl:761]: Let's call this Bolt12Offer?
https://github.com/breez/breez-sdk-greenlight/pull/1121#issuecomment-2468030319

Comment (ok300) [reply libs/sdk-bindings/src/breez_sdk.udl]: Done: a8a7057320aca4f5b03708a11de0ea66dc931d80

**Comment by ok300** [comment]
As discussed, this is adding a simple `Bolt12Offer { offer: String }` variant, until `gl-client` supports the newer `lightning` and `lightning-invoice` crates.

In addition, Liquid SDK will attempt to parse the offer locally.

File: libs/sdk-bindings/src/breez_sdk.udl (hunk 1 above)
Comment (ok300) [reply libs/sdk-bindings/src/breez_sdk.udl]: Addressed: https://github.com/breez/breez-sdk-greenlight/pull/1121#issuecomment-2472855649

File: libs/sdk-common/src/invoice.rs [hunk 3]
```diff
@@ -133,8 +133,10 @@ impl From<lightning::offers::offer::Amount> for Amount {
 
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
 pub struct LNOffer {
+    /// String representation of the Bolt12 offer
     pub bolt12: String,
```
Comment (dangeross) [inline libs/sdk-common/src/invoice.rs:137]: Maybe `offer` is more understandable?

File: libs/sdk-common/src/invoice.rs [hunk 4]
```diff
@@ -101,6 +101,50 @@ fn format_short_channel_id(id: u64) -> String {
...
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub enum Amount {
+    Bitcoin {
+        amount_msat: u64,
+    },
+    Currency {
+        // Reference to [FiatCurrency.id]
+        iso4217_code: String,
```
Comment (dangeross) [inline libs/sdk-common/src/invoice.rs:111]: Maybe a name that more obviously links it to FiatCurrency? something like `currency_id`

File: libs/sdk-common/src/invoice.rs [hunk 5]
```diff
@@ -101,6 +101,50 @@ fn format_short_channel_id(id: u64) -> String {
...
+        fractional_amount: u64,
```
Comment (dangeross) [inline libs/sdk-common/src/invoice.rs:112]: Would this be like Euro cents? and you can use the FiatCurrency info to calculate the decimal equivalent

File: libs/sdk-common/src/invoice.rs [hunk 6]
```diff
@@ -101,6 +101,50 @@ fn format_short_channel_id(id: u64) -> String {
...
+        match amount {
+            lightning::offers::offer::Amount::Bitcoin { amount_msats } => Amount::Bitcoin {
+                amount_msat: amount_msats,
+            },
+            lightning::offers::offer::Amount::Currency {
+                iso4217_code,
+                amount,
+            } => Amount::Currency {
+                iso4217_code: String::from_utf8(iso4217_code.to_vec())
+                    .expect("Expecting a valid ISO 4217 character sequence"),
```
Comment (dangeross) [inline libs/sdk-common/src/invoice.rs:127]: Better to do a TryFrom?

File: libs/sdk-common/src/invoice.rs (hunk 4 above)
Comment (ok300) [reply libs/sdk-common/src/invoice.rs:111]: This doesn't necessarily link it to `FiatCurrency`. It just happens they both use the same type of ID. `currency_id` might be obvious for us, but maybe not for others who aren't aware of `FiatCurrency`.

 IMO `iso4217_code` is clearer to the caller.

File: libs/sdk-common/src/invoice.rs (hunk 6 above)
Comment (ok300) [reply libs/sdk-common/src/invoice.rs]: Good idea, done: caf483b6eebf860e079c239858f2900c82302955

File: libs/sdk-common/src/invoice.rs (hunk 3 above)
Comment (ok300) [reply libs/sdk-common/src/invoice.rs]: Done: 643928e4cab2aa0a77f9df5b1c202706181fe293

**Review by dangeross — APPROVED** [review]
LGTM

**Review by hydra-yse — COMMENTED** [review]
Should we also add parsing for `recurrence-` and `quantity` based fields, or are those outside the scope of what will be used? Perhaps they can be added later as they seem useless for now. Besides that, looking good! :+1:

**Comment by ok300** [comment]
IMO it's best to add them when they will also be tested and used.

**Review by hydra-yse — APPROVED** [review]

**Pull request merged**